### ISSUE TITLE
Can't sign up to be a host?

### ISSUE THREAD

**Issue by mohanzhang** [TeaWithStrangers/tws-on-rails#505]
TL;DR: I think there is a bug as a result of this commit: https://github.com/TeaWithStrangers/tws-on-rails/commit/881d848a12d8a40178b044db8a2db63d92f55f49

Long "story" (which according to the site makes it 10x more likely that I get a quick reply from Ankit ;):

I just tried to click the "I'm Interested" button on `/hosting` but am not sure what is supposed to happen besides the button getting a new CSS style. The counter on my city page didn't update, so I went looking at the view (https://github.com/TeaWithStrangers/tws-on-rails/blob/master/app/views/static/hosting.html.erb#L67), I'm guessing that the intention was to run some JS, so I searched for the magic string "user-interest" which led me to this: https://github.com/TeaWithStrangers/tws-on-rails/blob/e3db4c29297f4a947312a0d1d5881d82f5bc5186/app/assets/javascripts/cities.js#L95

Then I opened up the Network tab in Chrome Inspector and found that the PATCH does indeed happen, and that response is `{"success":true}`. The params that get passed are `tws_interests[type]:true`, which is slightly weird but not unthinkable. Following the route into the controller, I see this: https://github.com/TeaWithStrangers/tws-on-rails/blob/master/app/controllers/api/v1/users_controller.rb#L13 (this is going somewhere, I promise).

Finally, when I look at that `user.tws_interests` method, I see this: https://github.com/TeaWithStrangers/tws-on-rails/blob/master/app/models/user.rb#L100

I haven't tried running this locally, so I can't be too sure, but based on the behavior I'm seeing on the production site when I GET my interests (http://www.teawithstrangers.com/api/v1/users/self/interests and I see this: `{"hosting":false,"leading":false,"type":"true"}`), I think this line might be the culprit: https://github.com/TeaWithStrangers/tws-on-rails/blame/e3db4c29297f4a947312a0d1d5881d82f5bc5186/app/assets/javascripts/cities.js#L89

Courtesy perhaps of this jshint cleanup? https://github.com/TeaWithStrangers/tws-on-rails/commit/881d848a12d8a40178b044db8a2db63d92f55f49

Should be easy to fix with a revert of that commit for just that code block. Would be happy to put together pull request if you agree. Cheers!

**Comment by mehulkar** [comment]
@mohanzhang! you are awesome. Thanks for taking the time to track this down. 

Couple things:

1. [This line][1] in the ajax that expresses interest adds the data to the PATCH we want. I haven't validated that this is actually what the API expects, but I don't _think_ the refactor was the culprit. The commented out line you pointed out doesn't get used (should be removed)
2. Yes, if user's interests are not getting updated, this is bad and is a high priority bug. 
3. The current implementation of this whole feature seems a limiting, so it's possible that we should revisit all of it, but I'd be highly interested in a PR with fix. 

Let me know if you are planning to fix. If not, I'll put this on my TODO for tomorrow evening during my office hours.

[1]: https://github.com/TeaWithStrangers/tws-on-rails/blob/e3db4c29297f4a947312a0d1d5881d82f5bc5186/app/assets/javascripts/cities.js#L94

**Comment by mehulkar** [comment]
Couple things that look suspect in the code:

1. Rewriting a db value on the model level [here][1]. Seems like a bad idea and the API might just be returning the wrong thing.
2. Update interests endpoint returns success even if the update failed :cry: 

[1]: https://github.com/TeaWithStrangers/tws-on-rails/blob/e3db4c29297f4a947312a0d1d5881d82f5bc5186/app/assets/javascripts/cities.js#L94

**Comment by mohanzhang** [comment]
The reason I think the specific commit I linked is the culprit is because of [this portion](https://github.com/TeaWithStrangers/tws-on-rails/commit/881d848a12d8a40178b044db8a2db63d92f55f49#diff-21479949a4944e6020bdeaf3083fcb03L106) in the diff. The intention on the old line 108 was that `data['tws_interests'][type] = true` where type is assigned just a couple of lines earlier to be either "hosting" or "leading". In the post-cleaned up version, `type` is now a property on `data.tws_interests` that always gets set to true, which of course totally misses the whole point of `type` previously, which was to be either `data.tws_interests.hosting = true` or `data.tws_interests.leading = true`.

I think this will fix the bug immediately, but I also agree that a few more enhancements could make the experience a bit more user friendly. In addition to the things you mentioned, I was originally tripped up by the round-trip delay in seeing the button css update, so I thought that nothing happened and kept refreshing and trying again. There is also a missing `event.preventDefault()` somewhere too because my screen got bounced to the top on the empty anchor tag, which then added to my confusion (and contributed to my hypothesis that the page was refreshing when it in fact wasn't).

**Comment by mehulkar** [comment]
@mohanzhang great catch. that indeed is a bug. I know you can probably make the 1 line fix, but would you be able to write a test around this as well?
